components/support/rc_crypto/src/hmac.rs

```diff
@@ -0,0 +1,114 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{
+    constant_time, digest,
+    error::*,
+    p11,
+    util::{ensure_nss_initialized, map_nss_secstatus},
+};
+use std::ops::Deref;
+
+#[derive(Clone)]
+pub struct Signature(digest::Digest);
+
+impl AsRef<[u8]> for Signature {
+    #[inline]
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+pub struct SigningKey {
+    pub(crate) digest_alg: &'static digest::Algorithm,
+    pub(crate) key_value: Vec<u8>,
+}
+
+impl SigningKey {
+    #[inline(always)]
+    pub fn new(digest_alg: &'static digest::Algorithm, key_value: &[u8]) -> Self {
+        SigningKey {
+            digest_alg,
+            key_value: key_value.to_vec(),
+        }
+    }
+
+    #[inline]
+    pub fn digest_algorithm(&self) -> &'static digest::Algorithm {
+        self.digest_alg
+    }
+}
+
+pub struct VerificationKey {
+    wrapped: SigningKey,
+}
+
+impl VerificationKey {
+    #[inline(always)]
+    pub fn new(digest_alg: &'static digest::Algorithm, key_value: &[u8]) -> Self {
+        VerificationKey {
+            wrapped: SigningKey::new(digest_alg, key_value),
+        }
+    }
+
+    #[inline]
+    pub fn digest_algorithm(&self) -> &'static digest::Algorithm {
+        self.wrapped.digest_algorithm()
+    }
+}
+
+pub fn verify(key: &VerificationKey, data: &[u8], signature: &[u8]) -> Result<()> {
+    verify_with_own_key(&key.wrapped, data, signature)
+}
+
+pub fn verify_with_own_key(key: &SigningKey, data: &[u8], signature: &[u8]) -> Result<()> {
+    constant_time::verify_slices_are_equal(sign(key, data)?.as_ref(), signature)
+}
+
+pub fn sign(key: &SigningKey, data: &[u8]) -> Result<Signature> {
+    let mech = match key.digest_alg {
+        digest::Algorithm::SHA256 => nss_sys::CKM_SHA256_HMAC,
+    };
+    ensure_nss_initialized();
+    let sym_key = p11::import_sym_key(mech.into(), nss_sys::CKA_SIGN.into(), &key.key_value)?;
+    let context = p11::create_context_by_sym_key(mech.into(), nss_sys::CKA_SIGN.into(), &sym_key)?;
+    map_nss_secstatus(|| unsafe { nss_sys::PK11_DigestBegin(*context.deref()) })?;
+    map_nss_secstatus(|| unsafe {
+        nss_sys::PK11_DigestOp(*context.deref(), data.as_ptr(), data.len() as u32)
+    })?;
+    let mut out_len: u32 = 0;
+    let mut out = vec![0u8; nss_sys::HASH_LENGTH_MAX as usize];
```

Comment:
It may be worth a comment here about NSS wanting to return the `out_len`, even though we should be able to know it up-front based on the digest type.  It tooks me a litlte bit to figure out what was happening here.